apstools/utils.py

```diff
@@ -384,7 +393,377 @@ def itemizer(fmt, items):
+    except (ModuleNotFoundError, AttributeError):
+        ns = globals()
+    return ns
+
+
+@dataclass
+class ListRuns:
+    """
+    List the runs from the given catalog according to some options.
+
+    EXAMPLE::
+
+        ListRuns(cat).to_dataframe()
+
+    PUBLIC METHODS
+
+    .. autosummary::
+
+        ~to_dataframe
+        ~to_table
+        ~parse_runs
+
+    INTERNAL METHODS
+
+    .. autosummary::
+
+        ~_get_by_key
+        ~_check_cat
+        ~_apply_search_filters
+        ~_sorter
+        ~_check_keys
+
+    """
+
+    cat: object = None
+    query: dict = None
+    keys: str = None
+    missing: str = ""
+    num: int = 20
+    reverse: bool = True
+    since: str = None
+    sortby: str = "time"
+    timefmt: str = "%Y-%m-%d %H:%M:%S"
+    until: str = None
+
+    _default_keys = "scan_id time plan_name detectors"
+
+    def _get_by_key(self, md, key):
+        """
+        Get run's metadata value by key.
+
+        Look in ``start`` document first.
+        If not found, look in ``stop`` document.
+        If not found, report using ``self.missing``.
+
+        If ``key`` is found but value is None, report as ``self.missing``.
+
+        The ``time`` key will be formatted by the ``self.timefmt`` value.
+        See https://strftime.org/ for examples.  The special ``timefmt="raw"``
+        is used to report time as the raw value (floating point time as used in
+        python's ``time.time()``).
+
+        A special syntax of ``key`` allows reporting of keys in other
+        metadata subdictionaries.  The syntax is ``doc.key`` (such as
+        specifying the ``time`` key  from the ``stop`` document:
+        ``stop.time``) where a ``.`` is used to separate the
+        subdictionary name (``doc``) from the ``key``.
+        **Note**:
+        It is not possible to ``sortby`` the dotted-key syntax
+        at this time.
+        """
+        v = None
+        if key == "time":
+            v = md["start"][key]
+            if self.timefmt != "raw":
+                ts = datetime.datetime.fromtimestamp(v)
+                v = ts.strftime(self.timefmt)
+        elif key in md["start"]:
+            v = md["start"].get(key, self.missing)
+        elif md["stop"] and key in md["stop"]:
+            v = md["stop"].get(key, self.missing)
+        elif len(key.split(".")) == 2:
+            # dotted-key syntax
+            doc, key = key.split(".")
+            if md[doc] is not None:
+                v = md[doc].get(key, self.missing)
+                if key == "time" and self.timefmt != "raw":
+                    ts = datetime.datetime.fromtimestamp(v)
+                    v = ts.strftime(self.timefmt)
+        return v or self.missing
+
+    def _check_cat(self):
+        if self.cat is None:
+            self.cat = getCatalog()
+
+    def _apply_search_filters(self):
+        since = self.since or "1995-01-01"
+        until = self.until or "2100-12-31"
+        self._check_cat()
+        cat = self.cat.v2.search(
```

Comment:
one comment here...

One can actually merge the TimeRange and "regular" queries into something like:

```python
time_query = databroker.queries.TimeRange(since=since, until=until).query
merged_query = self.query.update(time_query) if self.query else time_query
cat = self.cat.v2.search(merged_query)
````

Now, it seems like this is rather useless for a MongoDB because it's very fast anyway:

```python
In [38]: from databroker import catalog; db = catalog['mongodb_config']

In [39]: %%time
    ...: time_query = TimeRange(until='2021-01-01')
    ...: query = dict(plan_name="scan")
    ...: query.update(time_query.query)
    ...: search1 = db.search(query)
    ...: 
    ...: 
CPU times: user 414 µs, sys: 47 µs, total: 461 µs
Wall time: 467 µs

In [40]: from databroker import catalog; db = catalog['mongodb_config']

In [41]: %%time
    ...: time_query = TimeRange(until='2021-01-01')
    ...: query = dict(plan_name="scan")
    ...: search2 = db.search(time_query).search(query)
    ...: 
    ...: 
CPU times: user 675 µs, sys: 0 ns, total: 675 µs
Wall time: 683 µs
```

but it does make more of a difference if the `db` is backed by files (using a "unpacked" `db` in my laptop with 1042 runs):

```python
In [11]: from databroker import catalog; db = catalog['4idd-nov20']

In [12]: %%time
    ...: time_query = TimeRange(until='2020-12-03')
    ...: query = dict(plan_name="scan")
    ...: query.update(time_query.query)
    ...: search1 = db.search(query)
    ...: 
    ...: 
CPU times: user 6.75 s, sys: 2.4 s, total: 9.15 s
Wall time: 12.6 s

In [13]: from databroker import catalog; db = catalog['4idd-nov20']

In [14]: %%time
    ...: time_query = TimeRange(until='2020-12-03')
    ...: query = dict(plan_name="scan")
    ...: search2 = db.search(time_query).search(query)
    ...: 
    ...: 
CPU times: user 13.2 s, sys: 4.69 s, total: 17.9 s
Wall time: 25.1 s
```

Reply:
Thanks for the analysis!